theano/sandbox/cuda/tests/test_opt.py

```diff
@@ -91,6 +92,34 @@ def test_local_gpu_contiguous_gpu_contiguous():
                      if isinstance(node.op, basic_ops.GpuContiguous)])
 
 
+def test_local_assert_no_cpu_op():
+    numpy.random.seed(1)
+    m = numpy.random.uniform(-1, 1, (10, 10)).astype("float32")
+    ms = cuda.shared_constructor(m, name="m_shared")
+    out = theano.tensor.tanh(ms**2 + 3)
+
+    mode_local_assert = mode_with_gpu.including("assert_no_cpu_op")
+    mode_local_assert = mode_local_assert.excluding("local_gpu_elemwise_1")
```

Comment:
In this particular case, I think it is actually `local_gpu_elemwise_0` that gets applied.
On top of that, since one of the two inputs to "add" is a constant, it is not considered a "gpu input", so it does not trigger the first part of the condition.
I used the following: `(ms**2 + ms).dot(ms.T)`, excluding both `local_gpu_elemwise_0` and `1`, to create a graph that has the case we want to detect, and realized that there indeed an error in the check.